cardano-crypto-tests/src/Test/Crypto/EllipticCurve.hs

```diff
@@ -0,0 +1,224 @@
+{-# OPTIONS_GHC -Wno-orphans #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE LambdaCase #-}
+
+module Test.Crypto.EllipticCurve
+where
+
+import qualified Cardano.Crypto.EllipticCurve.BLS12_381 as BLS
+import qualified Cardano.Crypto.EllipticCurve.BLS12_381.Internal as BLS
+import Test.Crypto.Instances ()
+import Test.QuickCheck (
+--   (=/=), 
+    (===), 
+    (==>), 
+    Arbitrary(..), 
+    Property,
+    suchThatMap,
+--   Gen, 
+--   forAllShow
+  )
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.QuickCheck (testProperty)
+import Test.Tasty.HUnit (testCase, assertBool, assertEqual)
+import Data.Proxy (Proxy (..))
+import qualified Data.ByteString as BS
+import System.IO.Unsafe (unsafePerformIO)
+import Data.Bits (shiftL)
+
+tests :: TestTree
+tests =
+  testGroup "Crypto.EllipticCurve"
+    [ testGroup "BLS12_381"
+        [ testUtil "Utility"
+        , testScalar "Scalar"
+        , testBLSCurve "Curve 1" (Proxy @BLS.Curve1)
+        , testBLSCurve "Curve 2" (Proxy @BLS.Curve2)
+        , testPT "PT"
+        , testPairing "Pairing"
+        ]
+    ]
+
+testUtil :: String -> TestTree
+testUtil name =
+  testGroup name
+    [ testProperty "Integer / C-String round-trip" $
+        \n ->
+          n >= 0 ==>
+          n === unsafePerformIO (BLS.integerAsCStr n BLS.cstrToInteger)
+    , testProperty "Integer / C-String 32 round-trip" $
+        \n ->
+          n >= 0 ==>
+          n < (1 `shiftL` 32 * 8) ==>
+          n === unsafePerformIO (BLS.integerAsCStrL 32 n BLS.cstrToInteger)
+    , testProperty "padBS min length" $ \n bsw ->
+        BS.length (BLS.padBS n (BS.pack bsw)) >= n
+    , testProperty "padBS adds zeroes to front" $ \bsw ->
+        BS.index (BLS.padBS (length bsw + 1) (BS.pack bsw)) 0 === 0
+    , testCase "integerToBS" $ do
+        assertEqual "0x1234" (BS.pack [0x12, 0x34]) (BLS.integerToBS 0x1234)
+        assertEqual "0x12345678" (BS.pack [0x12, 0x34, 0x56, 0x78]) (BLS.integerToBS 0x12345678)
+    ]
+
+testScalar :: String -> TestTree
+testScalar name =
+  testGroup name
+    [ testProperty "self-equality" $
+        \(a :: BLS.Scalar) -> a === a
+    , testProperty "to/from BS round-trip" $
+        \s -> Right s === (BLS.scalarFromBS . BLS.scalarToBS $ s)
+    , testProperty "non-negative" $
+        \s -> (unsafePerformIO . BLS.scalarToInteger $ s) >= 0
+    , testProperty "to/from Integer round-trip" $
+        \s -> s === ((unsafePerformIO . BLS.scalarFromInteger) . (unsafePerformIO . BLS.scalarToInteger) $ s)
+    , testCase "integer from scalar" $ do
+        s <- case BLS.scalarFromBS (BS.pack [0x12, 0x34]) of
+              Left err -> error (show err)
+              Right x -> return x
+        let expected = 0x1234
+        actual <- BLS.scalarToInteger s
+        assertEqual "0x1234" expected actual
+    ]
```

Comment:
I think we need to have a similar test tree for the Natural conversion functions

- Natural / C-String round-trip (`natAsCStr` / `cstrToNat`)
    - There's no `natFromBS` function, but the Natural / C-String round-trip uses `natToBS` so I guess that `natToBS` is tested here.
- Natural / C-String 32 round-trip (`natAsCStrL` / `cstrToNat`)
- Scalar / Natural round-trip (`scalarToNat` / `scalarFromNat`) (note `scalarFromNat` truncates 32 bytes)

Reply:
Actually, the entire C.hs module, and with it the conversions to and from `Nat`, are obsolete; it was already excluded from the .cabal file, but I have now also deleted it from the repo. In practice, we will only need the conversions to and from `Integer`.